apps/rush-lib/src/logic/buildCache/AzureStorageBuildCacheProvider.ts

```diff
@@ -0,0 +1,378 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { BuildCacheProviderBase, IBuildCacheProviderBaseOptions } from './BuildCacheProviderBase';
+import { Terminal } from '@rushstack/node-core-library';
+import {
+  BlobClient,
+  BlobSASPermissions,
+  BlobServiceClient,
+  BlockBlobClient,
+  ContainerClient,
+  generateBlobSASQueryParameters,
+  SASQueryParameters,
+  ServiceGetUserDelegationKeyResponse,
+  UserDelegationKey
+} from '@azure/storage-blob';
+import { DeviceCodeCredential, DeviceCodeInfo } from '@azure/identity';
+
+import { EnvironmentConfiguration, EnvironmentVariableNames } from '../../api/EnvironmentConfiguration';
+import { RushGlobalFolder } from '../../api/RushGlobalFolder';
+import {
+  BuildCacheProviderCredentialCache,
+  IBuildCacheProviderCredentialCacheEntry
+} from './BuildCacheProviderCredentialCache';
+import { URLSearchParams } from 'url';
+import { RushConstants } from '../RushConstants';
+import { Utilities } from '../../utilities/Utilities';
+
+export type AzureEnvironmentNames =
+  | 'AzureCloud'
+  | 'AzureChinaCloud'
+  | 'AzureUSGovernment'
+  | 'AzureGermanCloud';
+export enum AzureEnvironment {
+  AzureCloud,
+  AzureChinaCloud,
+  AzureUSGovernment,
+  AzureGermanCloud
+}
+
+export interface IAzureStorageBuildCacheProviderOptions extends IBuildCacheProviderBaseOptions {
+  storageContainerName: string;
+  storageAccountName: string;
+  azureEnvironment?: AzureEnvironment;
+  blobPrefix?: string;
+  isCacheWriteAllowed: boolean;
+  rushGlobalFolder: RushGlobalFolder;
+}
+
+const SAS_TTL: number = 7 * 24 * 60 * 60 * 1000; // Seven days
```

Comment:
What happens to the CI machine after 7 days?  How does it refresh its token?

Reply:
Presumably you'll want to generate a new SAS and stick it in an env variable at the beginning of each run.